pylint/plugins/hass_enforce_type_hints.py

```diff
@@ -2534,18 +2536,41 @@ def _is_valid_return_type(match: TypeHintMatch, node: nodes.NodeNG) -> bool:
 
     if (
         match.check_return_type_inheritance
-        and isinstance(match.return_type, str)
+        and isinstance(match.return_type, (str, list))
         and isinstance(node, nodes.Name)
     ):
-        ancestor: nodes.ClassDef
-        for infer_node in node.infer():
-            if isinstance(infer_node, nodes.ClassDef):
-                if infer_node.name == match.return_type:
+        if isinstance(match.return_type, str):
+            valid_types = {match.return_type}
+        elif isinstance(match.return_type, list):
+            valid_types = {el for el in match.return_type if isinstance(el, str)}
+        else:
+            pass
+
+        try:
+            for infer_node in node.infer():
+                if _check_ancestry(infer_node, valid_types):
                     return True
-                for ancestor in infer_node.ancestors():
-                    if ancestor.name == match.return_type:
-                        return True
+        except NameInferenceError:
+            for class_node in node.root().get_children():
+                if (
+                    isinstance(class_node, nodes.ClassDef)
+                    and class_node.name == node.name
+                ):
+                    for infer_node in class_node.infer():
+                        if _check_ancestry(infer_node, valid_types):
+                            return True
+
+    return False
 
+
+def _check_ancestry(infer_node: nodes.FunctionDef, valid_types: set[str]) -> bool:
```

Comment:
`node.infer()` will return a Generator with type `InferenceResult` (from `astroid.typing`). I.e. `infer_node` should be typed with `InferenceResult`.

Reply:
I had to update Astroid for the type hint (I was on astroid 2.11.6)
Hopefully it won't cause too many issues for users with older astroid version, or do we need to add a check somewhere?

Reply:
I just pinned `2.12.5` in the requirements file yesterday (due to another issue) -> #77862.
However, I do get your point. Especially for those that don't update their venvs regularly. If you move the import behind `if TYPE_CHECKING` it will continue to work even with older versions.

Reply:
Yep - tested and working.